cmd/goat/plc.go

```diff
@@ -320,3 +417,344 @@ func fetchPLCData(ctx context.Context, plcHost string, did syntax.DID) (*PLCData
+	}
+
+	fmt.Println(didplc)
+
+	return nil
+}
+
+func runPLCSign(cctx *cli.Context) error {
+	s := cctx.Args().First()
+	if s == "" {
+		return fmt.Errorf("need to provide PLC operation json path as input")
+	}
+
+	privStr := cctx.String("plc-private-rotation-key")
+	if privStr == "" {
+		return fmt.Errorf("private key must be provided")
+	}
+
+	inputReader, err := getFileOrStdin(s)
+	if err != nil {
+		return err
+	}
+
+	inBytes, err := io.ReadAll(inputReader)
+	if err != nil {
+		return err
+	}
+
+	var enum didplc.OpEnum
+	if err := json.Unmarshal(inBytes, &enum); err != nil {
+		return err
+	}
+	op := enum.AsOperation()
+
+	// Note: we do not require that the op is currently unsigned.
+	// If it's already signed, we'll re-sign it.
+
+	privkey, err := crypto.ParsePrivateMultibase(privStr)
+	if err != nil {
+		return err
+	}
+
+	if err := op.Sign(privkey); err != nil {
+		return err
+	}
+
+	res, err := json.MarshalIndent(op, "", "  ")
+	if err != nil {
+		return err
+	}
+	fmt.Println(string(res))
+
+	return nil
+}
+
+func runPLCSubmit(cctx *cli.Context) error {
+	ctx := context.Background()
+	expectGenesis := cctx.Bool("genesis")
+	didString := cctx.String("did")
+
+	if !expectGenesis && didString == "" {
+		return fmt.Errorf("exactly one of either --genesis or --did must be specified")
+	}
+
+	if expectGenesis && didString != "" {
+		return fmt.Errorf("exactly one of either --genesis or --did must be specified")
+	}
+
+	s := cctx.Args().First()
+	if s == "" {
+		return fmt.Errorf("need to provide PLC operation json path as input")
+	}
+
+	inputReader, err := getFileOrStdin(s)
+	if err != nil {
+		return err
+	}
+
+	inBytes, err := io.ReadAll(inputReader)
+	if err != nil {
+		return err
+	}
+
+	var enum didplc.OpEnum
+	if err := json.Unmarshal(inBytes, &enum); err != nil {
+		return err
+	}
+	op := enum.AsOperation()
+
+	if op.IsGenesis() != expectGenesis {
+		if expectGenesis {
+			return fmt.Errorf("expected genesis operation, but a non-genesis operation was provided")
+		} else {
+			return fmt.Errorf("expected non-genesis operation, but a genesis operation was provided")
+		}
+	}
+
+	if op.IsGenesis() {
+		didString, err = op.DID()
+		if err != nil {
+			return err
+		}
+	}
+
+	if !op.IsSigned() {
+		return fmt.Errorf("operation must be signed")
+	}
+
+	c := didplc.Client{
+		DirectoryURL: cctx.String("plc-host"),
+		UserAgent:    GOAT_PLC_USER_AGENT,
+	}
+
+	if err = c.Submit(ctx, didString, op); err != nil {
+		return err
+	}
+
+	// TODO: print confirmation?
+
+	return nil
+}
+
+// fetch logs from /log/audit, select according to base_cid ("" means use latest), and
+// prepare it for updates:
+//   - convert from legacy op format if needed (and reject tombstone ops)
+//   - strip signature
+//   - set `prev` to appropriate value
+func fetchOpForUpdate(ctx context.Context, c didplc.Client, did string, base_cid string) (*didplc.RegularOp, error) {
+	auditlog, err := c.OpLog(ctx, did, true) // NB: this API changes in a pending go-didplc PR
```

Comment:
Will get the `go-didplc` changes merged before landing this PR, maybe a v0.0.1 tag too